blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md

```diff
@@ -0,0 +1,239 @@
+lines: `-`, `+` then `*` (and then multi-character bullets,
+I guess). You can also mix bullets with braces: the opening brace
+resets the bullet scope, any bullet can be used again with the
+subgoal.
+
+This gives a large space of freedom in how you want to use these
+features. You can use only braces, only bullets, braces and only one
+level of bullets, etc. My own style evolved with experience using the
+feature, and I will present the current status below.
+
+# My current bullet style
+
+When deciding how to use bullets, one distinguishes the commands that
+preserve the number of subgoals and those that may create new
+subgoals. I do some further distinctions.
+
+Some tactics, for example `assert`, create a number of subgoals that
+is *statically* known, always the same for the tactic. I then use
+braces around each sub-proof, except the last one, which I think of as
+the "rest" of the current proof.
+
+```
+assert foo as H.
+{ proof of foo. }
+rest of the proof using H:foo.
+```
+
+(If the proof of `foo` takes several lines, I two-indent them, with
+the braces alone on their lines.)
+
+Most tactics create a *dynamic* number of subgoals, that depends on
+the specifics of the objects being operated on; this is the case of
+`case`, `destruct`, `induction` for example. In this case, I open
+a brace before the tactic, and use a bullet for each subgoal.
+
+```
+{ induction foo; simpl; auto.
+- proof of first remaining subgoal.
+- proof of second remaining subgoal.
+  rest of the proof of the second subgoal.
+}
+```
+
+(Notice that the subgoal-creating step is vertically aligned with the
+proof steps: I use both braces and bullets, but take only one
+indentation level each time.)
+
+As an exception, I may omit the braces if we are at the toplevel of
+the proof (`Proof .. Qed` serve as braces).
+
+Note that omitting the braces here and using different bullets when
+you nest is also just fine. In my experience it gives proofs that are
+a bit more pleasant to read but also a bit more cumbersome to edit and
+move around.
+
+Finally, a not-uncommon mode of use of "dynamic" tactics in the sense
+above is to expect all the cases, except one, to be discharged by
+direct automation (for example they are all absurd except one). When
+it is my intent that all cases but one be discharged (and not
+a coincidence), I express it by not using braces (this command
+preserves the number of subgoals), but marking the remaining subgoal
+with a new bullet *without* increasing the indentation level.
+
+```
+{ induction foo.
+- first subgoal.
+- second subgoal.
+  case blah; discharge all sub-subgoals but one.
++ remaining sub-subgoal of the second subgoal.
+  finish the sub-subgoal.
+- third subgoal.
+}
+```
+
+(This is the only time where I use several bullet levels.)
+
+If you are the kind of programmer that is passionate about indentation
+style, I should now have tricked you to use bullets to propose
```

Comment:
Don't really follow this sentence, why would someone interested in indentation style want to use bullets?